### PR TITLE
feat: add support for dynamic data for combobox and multiselect

### LINKED ISSUE #454: Add support for dynamic data to Select components
### What I'm trying to achieve

I want to add the ability to inform the end users they need to load additional data for `Select`, `Combobox`, and `Muliselect` components.

### Describe a proposed solution

Components should expose callbacks that can then be used by the end user to fetch additional data.

| Component | Callbacks |
|--------|--------|
| `Select` | `onScroll` |
| `Combobox` | `onScroll`, `onAutocomplete` |
| `Multiselect` | `onScroll`, `onAutocomplete` |

### PR DESCRIPTION
**Pull request by krzysztofzuraw**
I want to merge this change because this PR supports dynamic options in the `Multiselect`, and `Combobbox` components. Unfortunately, I have to change API and introduce a breaking change. From this PR value is passed to the `Select`, `Multiselect`, and `Combobox` components should be in the form of `{value: string, label: string}` or `Array<{value: string, label: string}>` instead of `string | string[]` as before.

This change is needed to support dynamic values in mainly the `Multiselect` component but I ported it back to other components so we can have a common API.

See the table below with API changes:

<table>
<tr>
<td> Before </td> <td> After </td>
</tr>
<tr>
<td>

```tsx
<Select 
  value="value-1" 
  options={[{ label: "label-1", value: "value-1" }]} 
/>
```

</td>
<td>
    
```tsx
<Select 
  value={{label: 'label-1', value: 'value-1'}}
  options={[{label: 'label-1', value: 'value-1'}]} 
/>
```
</td>
</tr>
</table>

This PR closes #454

### Pull Request Checklist

- [ ] New components are exported from `./src/components/index.ts`.
- [ ] Storybook story is created and documentation properly generated.

### CODE REVIEW

File: src/components/BaseSelect/LoadingListItem/Spinner.tsx [hunk 1]
```diff
@@ -0,0 +1,20 @@
+import { spinner } from "./Spinner.css";
+
+export const Spinner = () => (
+  <svg
+    xmlns="http://www.w3.org/2000/svg"
+    width="16"
+    height="16"
+    fill="none"
+    className={spinner}
+  >
```
Comment (poulch) [inline src/components/BaseSelect/LoadingListItem/Spinner.tsx:10]: suggestion: you could use `createSVGWrapper` here

File: src/components/Combobox/Combobox.tsx [hunk 2]
```diff
@@ -31,9 +33,14 @@ export type ComboboxProps = PropsWithBox<
...
-    options: Option[];
+    options: ComboboxOption[];
     onChange?: ChangeHandler;
-    value: InputValue;
+    value: ComboboxOption | null;
+    onInputValueChange?: (value: string) => void;
+    loading?: boolean;
+    locale?: {
+      loading?: string;
+    };
```
Comment (poulch) [inline src/components/Combobox/Combobox.tsx:43]: question:  why not `loadingText`?

File: src/components/Multiselect/types.ts [hunk 3]
```diff
@@ -0,0 +1,11 @@
+import { UseComboboxPropGetters } from "downshift7";
+import { ReactNode } from "react";
+
+export type MultiselectOption = { label: string; value: string };
```
Comment (poulch) [inline src/components/Multiselect/types.ts:4]: suggestion: IMO this is a little bit redundant, we should have singe type for example Option

File: src/components/Multiselect/useMultiselect.tsx [hunk 4]
```diff
@@ -140,10 +138,16 @@ export const useMultiselectEvents = (
...
-          onFocus,
-          onBlur,
+          onFocus: (e: FocusEvent<HTMLInputElement, Element>) => {
+            _onFocus();
+            onFocus?.(e);
+          },
+          onBlur: (e: FocusEvent<HTMLInputElement, Element>) => {
+            _onBlur();
+            onBlur?.(e);
+          },
```
Comment (poulch) [inline src/components/Multiselect/useMultiselect.tsx:150]:
```suggestion
          onFocus: (e: FocusEvent<HTMLInputElement, Element>) => {
            setActive(true)
            onFocus?.(e);
          },
          onBlur: (e: FocusEvent<HTMLInputElement, Element>) => {
           setActive(false)
            onBlur?.(e);
          },
```

File: src/utils/useIntersectionObserver.tsx [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+  callback?: () => void;
+  options?: IntersectionObserverInit;
+}) {
+  const { threshold = 1, root = null, rootMargin = "0%" } = options;
+  const ref = useRef(null);
+
+  useEffect(() => {
+    const node = ref?.current;
+
+    console.log(node);
```
Comment (poulch) [inline src/utils/useIntersectionObserver.tsx:16]: sugggestion: remove log

File: src/utils/useIntersectionObserver.tsx [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+    if (!node) {
+      return;
+    }
+
+    const observer = new IntersectionObserver(
+      ([entry]) => {
+        console.log(entry.isIntersecting);
```
Comment (poulch) [inline src/utils/useIntersectionObserver.tsx:24]: suggestion: remove log

**Review by poulch — APPROVED** [review]

File: src/components/Select/Select.stories.tsx [hunk 7]
```diff
@@ -144,3 +140,42 @@ const [value, setValue] = useState("color-black");
...
+    <Select
+      value={value}
+      onChange={(value) => setValue(value)}
+      options={options}
+      loading={loading}
+      onScrollEnd={() => {
+        setLoading((prev) => !prev);
+        setOptions((prev) => [
+          ...prev,
+          { value: "color-white", label: "White" },
```
Comment (poulch) [inline src/components/Select/Select.stories.tsx:174]: suggestion: can we simulate async action, currently I can not see loading text

File: src/components/Multiselect/Multiselect.tsx [hunk 8]
```diff
@@ -183,20 +203,17 @@ export const Multiselect = forwardRef<HTMLInputElement, MultiselectProps>(
                     <Text size={size}>{item.label}</Text>
                   </List.Item>
                 ))}
+              {loading && (
+                <LoadingListItem size={size}>{locale.loading}</LoadingListItem>
+              )}
```
Comment (poulch) [inline src/components/Multiselect/Multiselect.tsx:208]: question: Do we want show this only when there is any items in dropdown?

**Review by poulch — APPROVED** [review]
LGTM

**Review by andrzejewsky — APPROVED** [review]

File: src/components/Combobox/Static/StaticCombobox.stories.tsx [hunk 9]
```diff
@@ -1,7 +1,8 @@
+/* eslint-disable react-hooks/rules-of-hooks */
```
Comment (poulch) [inline src/components/Combobox/Static/StaticCombobox.stories.tsx:1]: suggestion: if we have one line I'd prefer disable this rule only for this line not for whole file

**Review by poulch — APPROVED** [review]

**Review by andrzejewsky — APPROVED** [review]

**Pull request merged**